src/ocean/meta/traits/Indirections.d

```diff
@@ -0,0 +1,172 @@
+
+    T and all of its subtypes, if any, are expected to be
+      - an atomic type or
+      - a dynamic or static array or
+      - a struct or a union.
+
+    Params:
+        T = type to check
+
+    Returns:
+        true if T or any of its subtypes is a multi-dimensional dynamic array or
+        false otherwise.
+
+*******************************************************************************/
+
+public template containsMultiDimensionalDynamicArrays ( T )
+{
+    const containsMultiDimensionalDynamicArrays =
+        ReduceType!(T, MultiDimArraysReducer);
+}
+
+///
+unittest
+{
+    static assert (!containsMultiDimensionalDynamicArrays!(int));
+    static assert ( containsMultiDimensionalDynamicArrays!(int[][]));
+
+    static struct S
+    {
+        int[][] arr;
+    }
+
+    static assert ( containsMultiDimensionalDynamicArrays!(S));
+}
+
+private struct MultiDimArraysReducer
+{
+    alias bool Result;
+    const seed = false;
+
+    static Result accumulate ( Result accum, Result next )
+    {
+        return accum || next;
+    }
+
+    static Result visit ( T ) ( )
+    {
+        static if (is(T U == U[]))
+            return isArrayType!(U) == ArrayKind.Dynamic;
+        else
+            return false;
+    }
+}
+
+unittest
+{
+    static assert(!containsMultiDimensionalDynamicArrays!(int));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[ ]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[3]));
+
+    static assert( containsMultiDimensionalDynamicArrays!(int[ ][ ]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[3][ ]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[ ][3]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[3][3]));
+
+    static assert( containsMultiDimensionalDynamicArrays!(int[ ][ ][ ]));
+    static assert( containsMultiDimensionalDynamicArrays!(int[3][ ][ ]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[ ][3][ ]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[3][3][ ]));
+    static assert( containsMultiDimensionalDynamicArrays!(int[ ][ ][3]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[3][ ][3]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[ ][3][3]));
+    static assert(!containsMultiDimensionalDynamicArrays!(int[3][3][3]));
+
+    static assert(!containsMultiDimensionalDynamicArrays!(void));
+    static assert(!containsMultiDimensionalDynamicArrays!(void[]));
+    static assert( containsMultiDimensionalDynamicArrays!(void[][]));
+    static assert(!containsMultiDimensionalDynamicArrays!(void[][3]));
+
+    struct A
+    {
+        int x;
+        char[] y;
+        float[][][3][] z;
+    }
+
+    struct B
+    {
+        A[] a;
+    }
+
+    static assert(containsMultiDimensionalDynamicArrays!(A));
+
+    struct C
+    {
+        int x;
+        float[][3][] y;
+        char[] z;
+    }
+
```

Comment:
otally should but I haven't added any tests of my own so far, only copied old ones. Will try adding some new